File: scala/src/main/scala/io/devorbit/sparklabs/batch/sales/rdd/OrdersPerCity.scala
package io.devorbit.sparklabs.batch.sales.rdd

object OrdersPerCity extends App {
  val customersById = RddUtils.customers.map(customer => (customer.customer_id, customer.city))
  val ordersByCustomer = RddUtils.orders.map(order => (order.customer_id, 1))

  customersById.join(ordersByCustomer)
    .map { case (_, (city, _)) => (city, 1) }
    .reduceByKey(_ + _)
    .sortBy(_._2, ascending = false)
    .collect()
    .foreach(println)
}

File: python/sparklabs/batch/sales/rdd/orders_per_city.py
from sparklabs.batch.sales.rdd import utils


def run() -> None:
    rows = (
        utils.customers.map(lambda customer: (customer["customer_id"], customer["city"]))
        .join(utils.orders.map(lambda order: (order["customer_id"], 1)))
        .map(lambda item: (item[1][0], 1))
        .reduceByKey(lambda left, right: left + right)
        .sortBy(lambda item: item[1], ascending=False)
        .collect()
    )
    for row in rows:
        print(row)


if __name__ == "__main__":
    run()
